WIP: vulkan pixel history

apazylbe opened the pull request:
This is still WIP, only includes copying of pre and post modification values

apazylbe committed:
WIP: vulkan pixel history

Copy pre and post modification values

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,607 @@
+
+    // TODO: add all logic for draw calls.
+
+    m_pDriver->GetRenderState() = m_PrevState;
```
Is this push/pop to `m_PrevState` necessary? I think any code we run in `CopyPixel` shouldn't change the stored RenderState, and the begin below will re-set everything necessary.

apazylbe replied:
Removed, initially where I had a TODO I had code that was modifying the state, so this was leftover from then

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,607 @@
+          (VkImage)m_pDriver->GetResourceManager()->GetCurrentResource(depthImage);
+      depthCopyParams.srcImageLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+      depthCopyParams.srcImageFormat = imginfo.format;
+      m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, depthCopyParams, offset + 16);
```
This 16 offset is liable to bit-rot if the `EventInfo` struct changes, it'd be better to adjust the defintion of `EventInfo` so that colour and depth for premod/postmod are always stored together in a sub-struct, and then use `offsetof` on that sub-struct.

apazylbe replied:
Done, thanks!

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,607 @@
+    m_pDriver->GetRenderState().BeginRenderPassAndApplyState(cmd, VulkanRenderState::BindGraphics);
+
+    // Get post-modification values
+    m_EventIndex++;
```
I think this assumes that the callback always calls PreDraw/PostDraw for events in order, which is not necessarily the case. The user can record command buffers A and B in that order, but then submit them in the order B then A. That will cause the callback to hit higher EIDs before lower EIDs.

The solution can either be to get the index in `m_Events` of the EID and use that for the index, or else store the actual order of events seen and use that as a remapping later.

apazylbe replied:
Changed.

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,607 @@
+  void PreEndCommandBuffer(VkCommandBuffer cmd) {}
+  void AliasEvent(uint32_t primary, uint32_t alias)
+  {
+    // don't care
```
We should care about aliases, and it poses a particular problem for pixel history - the aliased call could easily have different pre/post modification values but we have no infrastructure to record a different copy. I'm not sure how best to solve this - pixel history is likely going to need some ability to record each submitted instance of a command buffer separately with a different copy, or else to do an indirected copy via vkCmdDispatch() that can then be re-based between each submission.

This also ties slightly into the above concern about secondary command buffers.

apazylbe replied:
Thanks for the feedback ! I added a TODO for now.

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,607 @@
+
+    m_pDriver->GetRenderState() = m_PrevState;
+    if(m_PrevState.graphics.pipeline != ResourceId())
+      m_pDriver->GetRenderState().BeginRenderPassAndApplyState(cmd, VulkanRenderState::BindGraphics);
```
This poses a problem with secondary command buffers - we can't stop or start render passes within them.

The only solution I can think of is that we'll need to add some ability to flatten out secondary command buffers into the primary command buffer and record directly there, so that we can split the render pass.

apazylbe replied:
Added a TODO for now.
If we do that, would that have to be done on capture side of replay?

apazylbe replied:
edit: capture side OR replay

baldurk replied:
It would have to be replay time, trying to flatten at capture time would be too intensive and bug-prone, and I wouldn't want to make such a drastic change to the command stream affecting everything if it's only needed for pixel history.

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,607 @@
+      m_pDriver, x, y, (VkImage)GetResourceManager()->GetCurrentResource(target), imginfo.format,
+      sampleMask, occlusionPool, resources.colorImageView, resources.stencilImageView,
+      resources.colorImage, resources.stencilImage, resources.dstBuffer, events);
+  m_pDriver->ReplayLog(events.front().eventId, events.back().eventId, eReplay_Full);
```
Since the callback already ignores other events than the ones it cares about, I'd suggest making this replay from event 0, and removing the above 'replay to first event'.

Generally speaking replaying from event X to event Y is not supported on vulkan especially when it can cross command buffer or submission boundaries. There's limited support when it's within the same command buffer but that won't always be the case here.

apazylbe replied:
Done.

baldurk reviewed (changes requested):
I've left high-level feedback. I think this is going along the right direction but unfortunately Vulkan poses some unique challenges which will make things difficult to iterate across the frame and perform work per-draw like this.

I didn't go into smaller nit-picky level for now since the code might change at a higher level anyway.

apazylbe committed:
Address feeedback from review

baldurk commented on renderdoc/driver/vulkan/vk_debug.cpp:
```diff
@@ -1423,6 +1423,13 @@ uint32_t VulkanReplay::PickVertex(uint32_t eventId, int32_t w, int32_t h, const
   return ret;
 }
 
+VulkanCreationInfo::Image VulkanDebugManager::GetImageInfo(ResourceId img)
```
This return value would be better returning a const reference I think, rather than a copy.

apazylbe replied:
Done

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,604 @@
+#include <stdio.h>
+#include <sys/types.h>
```
I'm not sure `sys/types.h` exists portably on windows. What did you need it for, is there a portable alternative or one already available in `os_specific.h`? Similarly what is stdio needed for here?

The file is also missing the copyright comment header.

apazylbe replied:
Sorry, I should have removed stdio, for sys/types.h I am not sure why I put it there, I had a lot more code there that I deleted, so it might have been left from that. I removed it now.
Thanks for catching this!

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,604 @@
+    colourCopyParams.srcImageLayout =
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;    // TODO: image layout
+    colourCopyParams.srcImageFormat = m_Format;
+    colourCopyParams.imageOffset = VkOffset3D{int32_t(m_X), int32_t(m_Y), 0};
```
Prefer not using this modern C++ style with brace constructors. Constructing from a plain list of items is fine but if you invoke a constructor it should be with normal brackets.

apazylbe replied:
DOne

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,604 @@
+#include "driver/vulkan/vk_replay.h"
+#include "vk_shader_cache.h"
+
+bool isUsageUAV(ResourceUsage usage)
```
The term 'UAV' isn't correct for Vulkan, so this function should have a better name. Perhaps `isDirectWrite` to go along with the final variable stored in the pixel history results.

apazylbe replied:
Right, makes sense, changed.

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,604 @@
+      CopyPixelParams depthCopyParams = colourCopyParams;
+      depthCopyParams.depthcopy = true;
+      depthCopyParams.srcImage =
+          (VkImage)m_pDriver->GetResourceManager()->GetCurrentResource(depthImage);
```
For the compile error about this cast (and another I think) look at `GetCurrentHandle` instead which should do the trick.

apazylbe replied:
Oh nice, thank you!

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -0,0 +1,604 @@
+{
+  VkImageLayout imgLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+  const ImageLayouts &imgLayouts = m_pDriver->m_ImageLayouts[image];
+  for(const auto &resState : imgLayouts.subresourceStates)
```
Avoid using `auto` except where the type is verbosely ugly, like lambdas or iterators.

apazylbe replied:
Done

baldurk reviewed (changes requested):
I've made some smaller more detailed comments. It should be OK to merge this after those are fixed and CI is green, with the proviso that it's still WIP. Note - as in the contributing doc I ask that each individual commit complies with the formatting rules and compiles. If you want to keep the review feedback separate that's fine but you'd need to cherry-pick those fixes backwards at least.

One question - do you think we should put the implementation behind an `#if` temporarily? It's not currently accessible via the UI but users can invoke it with scripts. I'm not sure if that's a big concern, but it's worth considering since it might crash or return garbage. If you don't mind having a local change to switch `#if 0` to `#if 1` then it might be cleaner to prevent accidental use while you're working on it.

apazylbe committed:
Address comments

- add copyright, remove unnecessary includes
- remove auto
- add #if for pixel history implementation

apazylbe commented:
Thanks for the suggestion, I didn't realized this code can be accessed by the scripts. I added #if directive to guard.

baldurk commented:
Thanks I think this looks good now, aside from the commit formatting issues mentioned above. Would you like me to squash everything together into the first commit, or would you like to rebase to fix them up?

apazylbe commented:
Would you be able to squash everything together into the first commit? Sorry, I was assuming we'd do squash and rebase when merging, missed your comments above. In the future I can do a single commit or make sure each commit is good.

baldurk commented:
Yeh I have the squash option on github web disabled to avoid accidentally squashing PRs that should be kept separate. I'll squash with a manual merge that's no problem. In future yes please if each commit can be formatted and compiling that's best.

apazylbe commented:
Sounds good, will do.  Thank you very much!

baldurk commented:
I renamed the commit to `vulkan pixel history - part 1` since I have hooks to prevent me accidentally pushing WIP commits, hopefully that's OK.
